src/types/basic.rs

```diff
@@ -0,0 +1,160 @@
+#![allow(missing_docs)]
+
+use super::{AbstractSignature, CustomType, TypeTag};
+
+enum EqTypeImpl {
+    USize,
+}
+enum ClassicTypeImpl {
+    E(EqTypeImpl),
+    Graph(Box<AbstractSignature>),
+}
+enum AnyTypeImpl {
+    C(ClassicTypeImpl),
+}
+
+pub struct Eq(EqTypeImpl);
+pub struct Classic(ClassicTypeImpl);
+pub struct Any(AnyTypeImpl);
+
+mod sealed {
+    use super::{Any, Classic, Eq};
+    pub trait Sealed {}
+    impl Sealed for Any {}
+    impl Sealed for Classic {}
+    impl Sealed for Eq {}
+}
+pub trait TypeClass: sealed::Sealed {
+    const TAG: TypeTag;
```

Comment:
I think I might call this STATIC_TAG just to drive that home. (It's what Rust'll guarantee for us, but the actual runtime tag could be narrower.)

We might actually define `fn tag(&self)` (or `dynamic_tag`) here as well, although I suspect this isn't quite the right place for that